### DISCUSSION TITLE
Handling Status Responses

### DISCUSSION THREAD

**Discussion by theotterlord** [withastro/roadmap#546] [Proposal]
### Body

# Summary

A better solution to handling different responses in SSR.

# Background & Motivation

The current way of dealing with a status code that would lead to another page (e.g. 404, or 500) is to do one of two things.

- Use `Astro.redirect` to send the user to another page (e.g. `/404`)
- Return a response from the page's code fence, streaming the response body fetched from the same page you would redirect too.

The former can be confusing, and the latter verbose. We've had a few people asking how to do this in support threads & elsewhere. By introducing an API, or some way to concisely provide the correct response it would help developers using SSR better implement these types of responses.

# Goals

- Introduce a concise, well documented API and/or recipe for easily handling common response types (404, 500, etc).

# Example

An idea for a potential API solution would be an `Astro.status` method that creates a `Response` by fetching the appropriate endpoint (if `/[status].astro` is a widely used convention). As well as just accepting a number (as seen below), it could also take a response like object, or a route to stream the response from. Very similar to `Astro.redirect` in practice, but likely with more options.

```astro
---
import { getContent } from 'my-cms'

let Content = await getContent(Astro.request)
if (!Content) {
    return Astro.status(404)
}
---
<Content/>
```

**Comment by florian-lefebvre** [comment]
I believe you can now do `Astro.response.status = 404`

**Reply by delucis** [reply]
Last time I did this with `return new Response(null, { status: 404 })` which will return the 404 page. I believe just setting the status just continues to render the current route but with a 404 status.

**Reply by florian-lefebvre** [reply]
Ah yes I thought it was just about setting the status code :+1:
